client/js/main.js

```diff
@@ -23,10 +23,27 @@ function getValidDropTargets(widget) {
         continue;
 
     let isValid = true;
-    for(const key in t.get('dropTarget')) {
-      if(widget.get(key) != t.get('dropTarget')[key] && (key != 'type' || widget.get(key) != 'deck' || t.get('dropTarget')[key] != 'card')) {
-        isValid = false;
-        break;
+    if(Array.isArray(t.get('dropTarget'))) {
+      isValid = false;
+      for(let i = 0; i < (t.get('dropTarget')).length; i++) {
+        let isValidObject = true;
+        for(const key in t.get('dropTarget')[i]) {
+          if((t.get('dropTarget')[i])[key] != (widget.get(key))) {
```

Comment:
Shouldn't this test match the test for non-array drop targets? Even better might be to refactor the code so that the test occurs in one place.

As currently written, an array of dropTarget objects will fail to match type deck when type card is specified, but the non-array version does allow decks to match when type card is specified. Of course the game author could explicitly make separate entries for type card and type deck.

Reply:
Unless I'm misunderstanding the non-array test allows for when dropTarget is null, and the purpose is to allow decks and cards to still be dropped (by default) when dropTarget is not set.  So I would not think that is needed when the dropTarget is an array.

As far as refactoring the code to have only 1 test, I'll have to take a look and see if I can figure something out.  I'm not really sure offhand.

Reply:
The default drop target is `{"type":"card"}` but the code allows a widget with `"type":"deck"` to match anyway. This is true even if a custom drop target is provided.

For refactoring, I was thinking a more generalized function similar to the `toA(ids)` function used in widget.js.

```javascript
function toArray(argument) {
//wraps argument in an array unless it is already an array
  return Array.isArray(argument) ? argument : [ argument ] ;
}
```

Reply:
> The default drop target is `{"type":"card"}` but the code allows a widget with `"type":"deck"` to match anyway.
Oh I see.  So I broke that for arrays, and need to fix it.

> This is true even if a custom drop target is provided.
You're talking about a secondary dropTarget criteria along with `{"type":"card"}` right? 

> 
> For refactoring, I was thinking a more generalized function similar to the `toA(ids)` function used in widget.js.
> 
> ```js
> function toArray(argument) {
> //wraps argument in an array unless it is already an array
>   return Array.isArray(argument) ? argument : [ argument ] ;
> }
> ```
Yeah, that's a good idea.  I feel bad that didn't occur to me.  However, should I just import that function then?  I don't know what is the best way to do these things.

Reply:
toA(ids) in widget.js gave me the idea, but is too specific to be imported for use here. More likely we should have a helper function somewhere that gets used in both places.

Reply:
I think I have a working version, but one thing that occurred to me is that `"dropTarget": []`, which some people might confuse with `"dropTarget": {}` will now accept nothing vs accepting everything on the PR server. 

Part of me thinks it might be a nice feature to leave it that way.  However, if so, then this probably needs an update to the fileupdater to change  `"dropTarget": []` into `"dropTarget": {}` for older games, just in case.

Reply:
Good point. We could simply check the array length and set isValid to true if the array is empty. That would avoid any breakage. However, we use empty arrays with the owner property to match none, so letting an empty array match everything with dropTarget might not be the best plan.

Reply:
`"dropTarget": []` also currently matches nothing in the PR (nothing will drop), so it sounds like that is consistent with `"owner": []`.  So I guess the thing to do would be to to add something to the fileupdater that changes `"dropTarget": []` to `"dropTarget": {}` for old games just in case someone is currently using it.